nodeup/pkg/model/etcd_manager_tls.go

```diff
@@ -59,5 +70,87 @@ func (b *EtcdManagerTLSBuilder) Build(ctx *fi.ModelBuilderContext) error {
 		}
 	}
 
+	// We also dynamically generate the client keypair for apiserver
+	if err := b.buildKubeAPIServerKeypair(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *EtcdManagerTLSBuilder) buildKubeAPIServerKeypair() error {
+	etcdClientsCACertificate, err := b.KeyStore.FindCert("etcd-clients-ca")
+	if err != nil {
+		return err
+	}
+
+	etcdClientsCAPrivateKey, err := b.KeyStore.FindPrivateKey("etcd-clients-ca")
+	if err != nil {
+		return err
+	}
+
+	if etcdClientsCACertificate == nil {
+		glog.Errorf("unable to find etcd-clients-ca certificate, won't build key for apiserver")
+		return nil
+	}
+	if etcdClientsCAPrivateKey == nil {
+		glog.Errorf("unable to find etcd-clients-ca private key, won't build key for apiserver")
+		return nil
+	}
+
+	dir := "/etc/kubernetes/pki/kube-apiserver"
+
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("error creating directories %q: %v", dir, err)
+	}
+
+	{
+		p := filepath.Join(dir, "etcd-ca.crt")
+		certBytes := certutil.EncodeCertPEM(etcdClientsCACertificate.Certificate)
+		if err := ioutil.WriteFile(p, certBytes, 0644); err != nil {
+			return fmt.Errorf("error writing certificate key file %q: %v", p, err)
+		}
+	}
+
+	name := "etcd-client"
+
+	humanName := dir + "/" + name
+	privateKey, err := certutil.NewPrivateKey()
+	if err != nil {
+		return fmt.Errorf("unable to create private key %q: %v", humanName, err)
+	}
+	privateKeyBytes := certutil.EncodePrivateKeyPEM(privateKey)
+
+	certConfig := certutil.Config{
+		CommonName: "kube-apiserver",
+		Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+
+	signingKey, ok := etcdClientsCAPrivateKey.Key.(*rsa.PrivateKey)
+	if !ok {
+		return fmt.Errorf("etcd-clients-ca private key had unexpected type %T", etcdClientsCAPrivateKey.Key)
+	}
+
+	glog.Infof("signing certificate for %q", humanName)
+	cert, err := certutil.NewSignedCert(certConfig, privateKey, etcdClientsCACertificate.Certificate, signingKey)
+	if err != nil {
+		return fmt.Errorf("error signing certificate for %q: %v", humanName, err)
+	}
+
+	certBytes := certutil.EncodeCertPEM(cert)
+
+	{
+		p := filepath.Join(dir, name+".crt")
+		if err := ioutil.WriteFile(p, certBytes, 0644); err != nil {
+			return fmt.Errorf("error writing certificate key file %q: %v", p, err)
+		}
+	}
+
+	{
+		p := filepath.Join(dir, name+".key")
```

Comment:
nit: Looks a little duplicated to me if the only change to line 142 is the file suffix

Reply:
There's also the byte slice we're writing, the file mode, and the error message :-)  I think it would be hard to dedup this - though very open to suggestions!

Reply:
What about 

```
{
    path := filepath.Join(dir, name)
    {
		if err := ioutil.WriteFile(p+".crt", certBytes, 0644); err != nil {
			return fmt.Errorf("error writing certificate key file %q: %v", p+".crt", err)
		}
    }
}
```

Not tested but would keep the filepath consistent. Even though I hate how it's appended two times now...

Reply:
Sure!  Done :-)